src/nsrt_learning.py

```diff
@@ -248,8 +244,102 @@ def _learn_pnad_side_predicates(
+                                         _get_successors, _evaluate)
+    # Final PNADs are the last ones in the path.
+    pnads = list(path[-1])
+    # Recompute the datastores in the PNADs. We need to do this
+    # because now that we have side predicates, each transition may be
+    # assigned to *multiple* datastores.
+    _recompute_datastores_from_segments(segmented_trajs, pnads)
+    return pnads
+
+
+def _recompute_datastores_from_segments(
+        segmented_trajs: List[List[Segment]],
+        pnads: List[PartialNSRTAndDatastore]) -> List[Set[Tuple[int, int]]]:
+    for pnad in pnads:
+        pnad.datastore = []  # reset all PNAD datastores
+    all_indices: List[Set[Tuple[int, int]]] = [set() for _ in pnads]
+    for traj_idx, seg_traj in enumerate(segmented_trajs):
+        objects = set(seg_traj[0].states[0])
+        for segment_idx, segment in enumerate(seg_traj):
+            identifier = (traj_idx, segment_idx)
+            if segment.has_option():
+                segment_option = segment.get_option()
+                segment_param_option = segment_option.parent
+                segment_option_objs = tuple(segment_option.objects)
+            else:
+                segment_param_option = DummyOption.parent
+                segment_option_objs = tuple()
+            # Get ground operators given these objects and option objs.
+            for pnad_idx, pnad in enumerate(pnads):
+                param_opt, opt_vars = pnad.option_spec
+                if param_opt != segment_param_option:
+                    continue
+                isub = dict(zip(opt_vars, segment_option_objs))
+                # Consider adding this segment to each datastore.
+                for ground_op in utils.all_ground_operators_given_partial(
+                        pnad.op, objects, isub):
+                    # Check if preconditions hold.
+                    if not ground_op.preconditions.issubset(
+                            segment.init_atoms):
+                        continue
+                    # Check if effects match. Note that we're using the side
+                    # predicates semantics here!
+                    atoms = utils.apply_operator(ground_op, segment.init_atoms)
+                    if not atoms.issubset(segment.final_atoms):
+                        continue
+                    # This segment belongs in this datastore, so add it.
+                    # TODO: this is an edge case we need to fix by changing
+                    # ObjToVarSubs in the partition to VarToObjSubs.
+                    if len(set(ground_op.objects)) != len(ground_op.objects):
+                        continue
+                    sub = dict(zip(ground_op.objects, pnad.op.parameters))
+                    pnad.add_to_datastore((segment, sub),
+                                          check_consistency=False)
+                    all_indices[pnad_idx].add(identifier)
+    return all_indices
```

Comment:
@tomsilver reminder for myself to get rid of this return value, we aren't pruning anymore (it hurts, bc you might prune an op that's good for planning even if it fits all the same data)

Reply:
interesting... but we still want to allow for pruning at the end of the day, I think. we should just do it in the same way that we're sidelining?

Reply:
i.e., test whether removing the operator improves the score function

Reply:
i can try that